spock-core/src/main/java/org/spockframework/runtime/extension/builtin/RetryIterationInterceptor.java

```diff
@@ -74,26 +77,32 @@ public void intercept(IMethodInvocation invocation) throws Throwable {
   }
 
   private class IterationState {
-    private final ThreadLocal<Boolean> finalIteration = ThreadLocal.withInitial(() -> false);
-    private final ThreadLocal<List<Throwable>> throwables = new ThreadLocal<>();
-
-    void startIteration() {
-      throwables.set(new ArrayList<>());
+    private final Map<IterationInfo, AtomicBoolean> finalIteration = new ConcurrentHashMap<>();
+    private final Map<IterationInfo, List<Throwable>> throwables = new ConcurrentHashMap<>();
+
+    void resetFailures(IterationInfo iteration) {
+      this.throwables.computeIfPresent(iteration, (key, throwables) -> {
+        throwables.clear();
+        return throwables;
+      });
     }
 
-    void setRetryAttempt(int retryAttempt) {
-      finalIteration.set(retryAttempt == retry.count());
+    void setRetryAttempt(IterationInfo iteration, int retryAttempt) {
+      finalIteration
+        .computeIfAbsent(iteration, key -> new AtomicBoolean(false))
+        .set(retryAttempt == retry.count());
     }
 
-    void failIteration(Throwable failure) {
-      throwables.get().add(failure);
-      if (finalIteration.get()) {
-        throw new MultipleFailuresError("Retries exhausted", throwables.get());
+    void failIteration(IterationInfo iteration, Throwable failure) {
+      List<Throwable> throwables = this.throwables.computeIfAbsent(iteration, key -> new CopyOnWriteArrayList<>());
+      throwables.add(failure);
+      if (finalIteration.containsKey(iteration) && finalIteration.get(iteration).get()) {
+        throw new MultipleFailuresError("Retries exhausted", throwables);
       }
     }
 
-    boolean notFailed() {
-      return throwables.get().isEmpty();
+    boolean notFailed(IterationInfo iteration) {
+      return !throwables.containsKey(iteration) || throwables.get(iteration).isEmpty();
```

Comment:
The following would save one map lookup and also makes it threadsafe, if throwables map would be cleared (which does currently not happen).

```suggestion
      List<Throwable> list = throwables.get(iteration);
      return list == null || list.isEmpty();
```